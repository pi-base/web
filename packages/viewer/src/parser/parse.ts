import { Parser } from '@pi-base/core'

import link, { Linkers } from './link'
import truncate from './truncate'

export type Options = {
  linkers?: Linkers
}

export function parser({ linkers = {} }: Options) {
  /* eslint-disable */
  // @ts-ignore
  const parser = () => Parser.html().use(link(linkers))

  return async function parse(body: string, truncated = false) {
    const p = truncated ? parser().use(truncate) : parser()
    return Parser.renderHTML(p, body)
  }
  /* eslint-enable */
}
