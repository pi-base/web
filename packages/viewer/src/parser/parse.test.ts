import { parser } from './parse'
import citation from './externalLinks'

function internalLink({ to }: { to?: string }) {
  if (!to) {
    return
  }
  if (to === 'missing') {
    return 'Could not find object'
  }

  return { href: `/objects/${to}`, label: to }
}

describe('parse', () => {
  const linkers = {
    citation,
    internalLink,
  }

  describe('untruncated', () => {
    const parse = parser({ linkers })

    test.skip('doi links', () =>
      expect(parse('{{doi:123}}')).resolves.toEqual(
        '<a href="https://doi.org/123">DOI 123</a>',
      ))

    test.skip('block math', async () => {
      const math = await parse('Thus $$2 + 2 = 4$$.')

      expect(math).toContain('Thus <span class="math-display">')
      expect(math).toMatchSnapshot()
    })

    test.skip('inline math', async () => {
      const math = await parse('Thus $2 + 2 = 4$.')

      expect(math).toContain('Thus <span class="math-inline">')
      expect(math).toMatchSnapshot()
    })

    test('unmatched', async () => {
      const math = await parse('A $ B')

      expect(math).toEqual('A $ B')
    })

    test.skip('internal links', () =>
      expect(parse('C.f. {S123}')).resolves.toEqual(
        'C.f. <a href="/objects/S123">S123</a>',
      ))

    test.skip('missing internal links', () =>
      expect(parse('C.f. {missing}')).resolves.toEqual(
        'C.f. <code to="missing">Could not find object</code>',
      ))
  })

  describe('truncated', () => {
    const parse = parser({ linkers })

    test('trims extra elements', () =>
      expect(parse('a list\n* one\n* two\n* three', true)).resolves.toEqual(
        'a list',
      ))
  })
})
