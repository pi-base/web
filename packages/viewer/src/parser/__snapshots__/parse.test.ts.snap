// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse untruncated block math 1`] = `"Thus <span class=\\"math-display\\"><span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.72777em;vertical-align:-0.08333em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">4</span></span></span></span></span></span>."`;

exports[`parse untruncated inline math 1`] = `"Thus <span class=\\"math-inline\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.72777em;vertical-align:-0.08333em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">4</span></span></span></span></span>."`;
